sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -524,14 +635,150 @@ protected EventProcessor()
+
+                // If the processor is not running, than it was never started or has been stopped
+                // before the semaphore was acquired; there is no work to be done.
+
+                if (_runningProcessorTask == null)
+                {
+                    return;
+                }
+
+                // Request cancellation of the running processor task.
+
+                _runningProcessorCancellationSource?.Cancel();
+                _runningProcessorCancellationSource?.Dispose();
+                _runningProcessorCancellationSource = null;
+
+                // Allow processing to complete.  If there was a processing or load balancing error,
+                // awaiting the task is where it will be surfaced.  Be sure to preserve it so
+                // that it can be surfaced.
+
+                try
+                {
+                    if (async)
+                    {
+                        await _runningProcessorTask.ConfigureAwait(false);
+                    }
+                    else
+                    {
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
+                        _runningProcessorTask.GetAwaiter().GetResult();
+#pragma warning restore AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
+                    }
+                }
+                catch (Exception ex) when ((ex is OperationCanceledException) || (ex is TaskCanceledException))
+                {
+                    // This is expected as part of the normal flow; no action is needed.
+                }
+                catch (Exception ex)
+                {
+                    Logger.EventProcessingTaskError(Identifier, EventHubName, ConsumerGroup, ex.Message);
+                    processingException = ex;
+                }
+
+                // With the processing task having completed, perform the necessary cleanup of partition processing tasks
+                // and surrender ownership.
+
+                var stopPartitionProcessingTasks = ActivePartitionProcessors.Keys
+                    .Select(partitionId => StopProcessingPartitionAsync(partitionId, ProcessingStoppedReason.Shutdown, CancellationToken.None))
+                    .ToArray();
+
+                if (async)
+                {
+                    await Task.WhenAll(stopPartitionProcessingTasks).ConfigureAwait(false);
+                    await LoadBalancer.RelinquishOwnershipAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    Task.WaitAll(stopPartitionProcessingTasks);
+
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
+                    LoadBalancer.RelinquishOwnershipAsync(cancellationToken).GetAwaiter().GetResult();
```

Comment:
FWIW - This feels like another place where in cases where `async` is false we are still going to end up awaiting.  The larger issue tracking this is (https://github.com/Azure/azure-sdk-for-net/issues/10083).  I'm not sure what the implications here are - but it feels like we may end up having some gnarly threading issues?

Reply:
Same thought as the one above.  #10083 was opened in response to questions on #10065 to try and explain the change being made.  I agree that whatever the outcome of the inquiry above should be applied here as well.